compiler/optimizer/VPHandlers.cpp

```diff
@@ -2759,15 +2759,15 @@ TR::Node *constrainIaload(OMR::ValuePropagation *vp, TR::Node *node)
             }
          }
       }
-
+   int32_t frequency = vp->_curBlock->getGlobalNormalizedFrequency(vp->comp()->getFlowGraph());
```

Comment:
I do not think that exposing this JIT internal frequency information directly to the GC and other components is a good design. My reasoning is as follows:

1) `getGlobalNormalizedFrequency` is just the block frequency multiplied by 1 <= warm, 10 @ hot, 100 @ scorching - block frequencies are not globally comparable. The frequency returned for every block _within a compilation_ is normalized between MAX_COLD_BLOCK_COUNT + 1 (current 6) and 10000. The hottest block has frequency 10000 and everything else is scaled from there. This means that if you inline a method with a really hot loop the frequencies of all the other blocks will drop way down. This makes block frequencies only comparable within a compile

2) Block frequencies have to be maintained by optimizations as the control flow graph is manipulated. There are a number of situations where no good block frequency can be derived and we make a best guess. This guess is designed to facilitate subsequent optimization - feeding these 'fake' frequencies to the GC is likely to introduce new levels of non-determinism.

3) Direct exposure of this frequency also means that anyone modifying the JIT will now have to worry about unintended consequences in the GC which is not a complexity most developers or reviewers can be reasonably assumed to be aware of and handle.

4) There is the additional concern that some profiling is namely more reliable than others - for example JProfiling frequency information from profiled compilations is naturally more recent and trustworthy than frequency information the IProfiler which may be well out of date when we do recompilations etc.

The biggest thing I'm trying to work through is the following situation:
A calls B calls C calls D.
C contains a hot loop which calls D where a field is being loaded. If C inlines D then we can understand how important the load in D is and mark it as hot. If C does NOT inline D then we won't see the field load in D as being in a loop (D is compiled on its own) and so we won't mark it as hot. This is really a whole program analysis question which  is something the JIT is not good at answering.... I also think that the hotness of field should not just be the hotness of the block - a field `f` loaded 20 times in blocks of frequency 5000 should be hotter than a field `g` loaded in one block of frequency 10000 (`f` is access more frequently).

I need to ponder some more on how to address the above concerns, but I am not happy to accept this implementation as-is because of the reasons I outlined above.

Reply:
@jonoommen do you have any thoughts on addressing the above points? This is cool work that we should try to incorporate so don't want this to stall because it isn't quite right on the first attempt....

Reply:
@andrewcraik I have spoken to @amicic regarding these concerns and he will reply regarding this, I appreciate the detailed feedback.

Reply:
Sounds good - I just wanted to make sure my concerns weren't treated as me thinking this was not a worthy item. I think this is great stuff, we just need to find the right way to capture what the GC wants/needs without coupling and scaling concerns - I look forward to hearing your thoughts and will consider pondering myself to see what solutions I can think of.

Reply:
Well, I don't have much to say other than: I mostly understand and do agree with the concerns.
One that I think it's not a big issue (as much you might think) is the situation with multiple loads within a block. We mostly care for (the first) one in a tight group, since it's likely the only one that will pay penalty of cache/TLB miss for subsequent loads of the same field. So per block info is not that bad.
One part missing for better understanding is when we get block frequency info - is it only on recompiles and that's info since the previous compile (or from interpreter or since explicit instrumentation for scorching compiles)? We don't get this later occasionally called with more up-to-date info? I guess, this is somewhat what you tried to say in (4). If we could occasionally get this info even after the last compile, we (GC) could accumulate them (prior to normalization), and that could help with the problem of uncomparable units among different methods, right?

For those 'fake' frequencies, I hope they are not big a problem, since they should hopefully be corrected, in subsequent re-compiles? Getting incorrect info for cold objects should not be big deal.

The (not) inlining scenario - yes, this is a big one. Not sure what to say - you'd better know how common this is and what to do about it...

What we can partially help with non-determinism is do profiling against various benchmarks and identify hot fields in JCL (like HashMap item and key) and mark them hot (for GC purpose) from the beginning. It would be based on a field name, so in unlikely scenario it changes between releases we can adjust. Jonny already kind of did that, but we wanted to minimize any static knowledge and we mostly removed them (except for String's char array that don't to show up as hot ever)

Reply:
So multiple loads in a block should get the same info when queried. I don't think that doing this in VP is quite the right place - we aren't doing anything here that really needs VP so we should move this logic to somewhere more sensible (possibly even a new pass specifically for it), but I think we need to address the higher level semantic issues first before worrying about such technicalities.

> One part missing for better understanding is when we get block frequency info - is it only on recompiles and that's info since the previous compile (or from interpreter or since explicit instrumentation for scorching compiles)? We don't get this later occasionally called with more up-to-date info?

Every compile will set the frequencies on blocks - whether it is the first compile or a recompile. The source of the information we use for setting the frequencies varies. If the first compile is before the IProfiler has gathered 'enough' data (determined by amount of samples) then we will use a static structural frequency generator to cook up some frequencies. Most initial compiles will use IProfiler data to derive the block frequencies from the branch bias info. When we recompile or upgrade the method to a higher opt level the frequencies will still come from the IProfiler data as at the time when we first compiled the method (no new profiling will have been recorded). If we do a profiled compile (limited to a small subset of the hottest methods in compute bound workloads) then the profile information in the recompile after the profling compile will have come from software profiling inserted into the code. This is JProfiling and is more accurate and up to date, but it is not generated for the vast majority of methods.

The 'fake' frequencies commonly affect loops when we convert them to a canonical form and any transformations affecting critical CFG edges. They are not rare or uncommon. We usually make good guesses about the frequencies, but they are guesses and could be surprising to you.

> The (not) inlining scenario - yes, this is a big one. Not sure what to say - you'd better know how common this is and what to do about it...

I don't think an empirical study is the right solution to this problem - we need to design a more robust way of deriving these frequencies. Inlining is an optimization and the choice of what to inline is a finely balanced act that the compiler will do to maximize performance. The choices can and do change and we should try to isolate this instrumentation / guidance from the vagaries of what an optimization chooses to do.

Some static guidance for the JCL would almost certainly help, but we do need to think about how to derive the global facts you are interested in. Sadly, it is a global question you are trying to answer and the limited context of a JIT compile makes producing good answers necessarily hard... I will keep pondering.